pkg/context/mqttcontext.go

```diff
@@ -0,0 +1,256 @@
+	case *packets.ConnectPacket:
+		mqttCtx.connectPacket = packet
+		mqttCtx.packetType = MQTTConnect
+	case *packets.PublishPacket:
+		mqttCtx.publishPacket = packet
+		mqttCtx.packetType = MQTTPublish
+	case *packets.DisconnectPacket:
+		mqttCtx.disconnectPacket = packet
+		mqttCtx.packetType = MQTTDisconnect
+	default:
+		mqttCtx.packetType = MQTTOther
+	}
+	return mqttCtx
+}
+
+// Protocol return protocol of mqttContext
+func (ctx *mqttContext) Protocol() Protocol {
+	return MQTT
+}
+
+// Deadline return deadline of mqttContext
+func (ctx *mqttContext) Deadline() (time.Time, bool) {
+	return ctx.ctx.Deadline()
+}
+
+// Done return done chan of mqttContext
+func (ctx *mqttContext) Done() <-chan struct{} {
+	return ctx.ctx.Done()
+}
+
+// Err return error of mqttContext
+func (ctx *mqttContext) Err() error {
+	ctx.mu.RLock()
+	defer ctx.mu.RUnlock()
+	if ctx.err != nil {
+		return ctx.err
+	}
+	return ctx.ctx.Err()
+}
+
+// Value return value of mqttContext for given key
+func (ctx *mqttContext) Value(key interface{}) interface{} {
+	return ctx.ctx.Value(key)
+}
+
+// Client return mqttContext client
+func (ctx *mqttContext) Client() MQTTClient {
+	return ctx.client
+}
+
+// Cancel cancel mqttContext
+func (ctx *mqttContext) Cancel(err error) {
+	ctx.mu.Lock()
+	if !ctx.canceled() {
+		ctx.err = err
+		ctx.cancelFunc()
+	}
+	ctx.mu.Unlock()
+}
+
+func (ctx *mqttContext) canceled() bool {
+	return ctx.err != nil || ctx.ctx.Err() != nil
+}
+
+// Canceled return if mqttContext is canceled
+func (ctx *mqttContext) Canceled() bool {
+	ctx.mu.RLock()
+	defer ctx.mu.RUnlock()
+	return ctx.canceled()
+}
+
+// Duration return time duration since this context start
+func (ctx *mqttContext) Duration() time.Duration {
+	ctx.mu.RLock()
+	defer ctx.mu.RUnlock()
+	if ctx.endTime != nil {
+		return ctx.endTime.Sub(*ctx.startTime)
+	}
+	return time.Since(*ctx.startTime)
+}
+
+// Finish tell this context is finished
+func (ctx *mqttContext) Finish() {
+	ctx.mu.Lock()
+	endTime := time.Now()
+	ctx.endTime = &endTime
+	ctx.mu.Unlock()
+}
+
+func (ctx *mqttContext) PacketType() MQTTPacketType {
+	return ctx.packetType
+}
+
+func (ctx *mqttContext) ConnectPacket() *packets.ConnectPacket {
+	return ctx.connectPacket
+}
+
+func (ctx *mqttContext) PublishPacket() *packets.PublishPacket {
+	return ctx.publishPacket
+}
+
+func (ctx *mqttContext) DisconnectPacket() *packets.DisconnectPacket {
+	return ctx.disconnectPacket
+}
+
+func (ctx *mqttContext) SetDrop() {
+	atomic.StoreInt32(&ctx.drop, 1)
+}
+
+func (ctx *mqttContext) Drop() bool {
+	return atomic.LoadInt32(&ctx.drop) == 1
+}
+
+func (ctx *mqttContext) SetDisconnect() {
+	atomic.StoreInt32(&ctx.disconnect, 1)
+}
+
+func (ctx *mqttContext) Disconnect() bool {
+	return atomic.LoadInt32(&ctx.disconnect) == 1
+}
+
+func (ctx *mqttContext) SetEarlyStop() {
+	atomic.StoreInt32(&ctx.earlyStop, 1)
+}
+
+func (ctx *mqttContext) EarlyStop() bool {
+	return atomic.LoadInt32(&ctx.earlyStop) == 1
+}
```

Comment:
Yeah, depend on the filter implementation, it may have multiple goroutines call these method at same time. Use atomic is a little bit faster than locks...